jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/CachingHttpContentFactory.java

```diff
@@ -329,43 +328,55 @@ public CachedHttpContent(String key, HttpContent httpContent)
             boolean isValid = true;
 
             // Read the content into memory if the HttpContent does not already have a buffer.
-            RetainableByteBuffer buffer;
+            Content.Chunk chunk;
             ByteBuffer byteBuffer = httpContent.getByteBuffer();
             if (byteBuffer == null)
             {
-                buffer = _bufferPool.acquire((int)_contentLengthValue, _useDirectByteBuffers);
                 try
                 {
                     if (_contentLengthValue <= _maxCachedFileSize)
                     {
-                        try (ReadableByteChannel readableByteChannel = httpContent.getResource().newReadableByteChannel())
+                        Content.Source contentSource = IOResources.asContentSource(httpContent.getResource(), _bufferPool, (int)_contentLengthValue, _useDirectByteBuffers);
+                        chunk = contentSource.read();
```

Comment:
Is this code assuming that a single read into a `Chunk` returns all the bytes?
I would not make such assumption.

Reply:
It's a direct translation of the `ReadableByteChannel` equivalent, which I agree is wrong.
I think `IOResources` could offer an API which uses an aggregator to read a resource into a RetainableByteBuffer.
Let me do that.

Reply:
Done.